content/en/blog/2022/merbridge/index.md

```diff
@@ -0,0 +1,163 @@
+
+When processing the inbound traffic, if the destination address is not in the map, we will not do anything to the traffic to make the traffic being processed more flexibly and simply.
+
+Otherwise, the steps are the same as for outbound traffic.
+
+{{< image link="./3.png" caption="Processing Inbound Traffic" >}}
+
+### Same-node acceleration
+
+Theoretically, acceleration between Envoy sidecars on the same node can be achieved directly through inbound traffic processing. However, Envoy will raise an error when accessing the application of the current pod in this scenario.
+
+In Istio, Envoy accesses the application by using the current pod IP and port number. With the above scenario, we realized that the pod ip is existing in the `local_pod_ips` as well, and the traffic will be redirected to the pod IP and 15006 port again because it is the same address that the inbound traffic comes from. Redirecting to the same inbound address will cause infinite recursion loop.
+
+Here comes the question: are there any ways to get the IP address in the current namespace with eBPF?
+
+The answer is Yes! We have designed a feedback mechanism:
+
+When Envoy tries to establish the connection, we do redirect it to port 15006. However, in the `sockops` step, we will determine if the source IP and the destination IP are the same. If yes, it means the wrong request is sent, and we will discard this connection in the `sockops` process. In the meantime, the current `ProcessID` and `IP` information will be written into the `process_ip` map, to allow eBPF to support the correspondence between processes and IPs.
+
+When the next request is sent, the same process above will not be performed anymore. We will check directly from the `process_ip` map if the destination address is the same as the current IP address.
+
+{{< warning >}}
+Envoy will retry when the request fails, and this retry process will only occur once, which means the subsequent requests will be super fast.
+{{< /warning >}}
+
+{{< image link="./4.png" caption="Same-node acceleration" >}}
+
+### Connection relationship
+
+Before applying eBPF using Merbridge, the data-path between pods is like:
+
+{{< image link="./5.png" caption="Iptables's Data-path" >}}
+
+After applying Merbridge(eBPF), the outbound traffic will skip many kernel modules like the graph below to improve the performance:
+
+{{< image link="./6.png" caption="EBPF's Data-path" >}}
+
+If two pods are on the same machine, the connection can even be faster:
+
+{{< image link="./7.png" caption="EBPF's Datapath on The Same Machine" >}}
+
+As above, using eBPF to process the traffic on the machine can greatly reduce the number of processes performed in the kernel, and the communication quality can be improved as well.
+
+## Performance test
+
+{{< warning >}}
+Currently the tests below are basic, and not tested in production yet.
+{{< /warning >}}
+
+The graph below shows the overall latency after using eBPF instead of iptables (lower is better):
```

Comment:
Two questions here, what's your payload size in your testing and what performance tool did you use to do the test?

Reply:
Each request about 200 bytes. and use `wrk` as performance tool.